### PR TITLE
Storage protocol resolver cache beans

### LINKED ISSUE #606: Cache protocol resolver settings for Storage
as mentioned in #574

### PR DESCRIPTION
**Pull request by ChengyuanZhao**
fixes #606

### CODE REVIEW

File: spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageProtocolResolver.java [hunk 1]
```diff
@@ -56,6 +60,21 @@ public void postProcessBeanFactory(
 		this.beanFactory = beanFactory;
 	}
 
+	private Storage getStorage() {
```
Comment (joaoandremartins) [inline spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageProtocolResolver.java]: A getter that changes the state of the object?

File: spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageProtocolResolver.java [hunk 2]
```diff
@@ -56,6 +60,21 @@ public void postProcessBeanFactory(
...
+		if (this.storage == null) {
+			this.storage = this.beanFactory.getBean(Storage.class);
+		}
+		return this.storage;
+	}
+
+	private GoogleStorageProtocolResolverSettings getGoogleStorageProtocolResolverSettings() {
```
Comment (joaoandremartins) [inline spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageProtocolResolver.java]: Ditto.

File: spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageProtocolResolver.java [hunk 3]
```diff
@@ -69,7 +88,7 @@ public void setResourceLoader(ResourceLoader resourceLoader) {
 
 	private GoogleStorageProtocolResolverSettings getSettings() {
 		try {
-			return this.beanFactory.getBean(GoogleStorageProtocolResolverSettings.class);
+			return getGoogleStorageProtocolResolverSettings();
```
Comment (joaoandremartins) [inline spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageProtocolResolver.java]: Why not inlining those getters here?
e.g.,
```
if (this.googleStorageProtocolResolverSettings == null) {
this.googleStorageProtocolResolverSettings = this.beanFactory.getBean(GoogleStorageProtocolResolverSettings.class);
}
return this.googleStorageProtocolResolverSettings;
```

Even better, do it in `postProcessBeanFactory()`, after `this.beanFactory = beanFactory`.

File: spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageProtocolResolver.java (hunk 1 above)
Comment (artembilan) [reply spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageProtocolResolver.java]: This is `private` method, so harmless

File: spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageProtocolResolver.java (hunk 3 above)
Comment (ChengyuanZhao) [reply spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageProtocolResolver.java]: I had that originally , but it appears things are not all ready to be used by that point. Specifically the GoogleProtocolResolverSettings isn't correct to be retrieved from the beanFactory at that point and unit tests fail.

Comment (artembilan) [reply spring-cloud-gcp-storage/src/main/java/org/springframework/cloud/gcp/storage/GoogleStorageProtocolResolver.java]: We can't do that in the `postProcessBeanFactory()`. That is going to be *evil* called `early BeanFactory` access.
That's why we resolve these bean lazy-load manner.

**Pull request merged**